likelihood/cl_like/bacco.py

```diff
@@ -0,0 +1,335 @@
+        k = self.ks / pars['hubble']
+
+        # 10 redshifts, 15 combinations between bias params, and the ks (added b0)
+        num_comb = 15
+        pk2d_bacco = np.zeros((len(self.a_s), num_comb, len(self.ks)))
+
+        # compute the power for each redshift
+        # t1 = time.time()
+        for i in range(len(self.a_s)):
+            pars['expfactor'] = self.a_s[i]
+            # call the emulator of the nonlinear 15 lagrangian bias expansion terms, shape is (15, len(k))
+            k, pnn = lpt_emu.get_nonlinear_pnn(pars, k=k)
+            pk2d_bacco[i, :, :] = pnn
+
+        # convert the spit out result from (Mpc/h)^3 to Mpc^3 (bacco uses h units, but pyccl doesn't)
+        pk2d_bacco /= pars['hubble'] ** 3
+        self.bacco_table = pk2d_bacco
+
+    def get_pgg(self, Pnl, b11, b21, bs1, b12, b22, bs2,
+                bk21=None, bk22=None, Pgrad=None):
+        if self.bacco_table is None:
+            raise ValueError("Please initialise BACCO calculator")
+        # Clarification:
+        # BACCO uses the following expansion for the galaxy overdensity:
+        #   d_g = b1 d + b2 d2^2 + bs s^2 + bnabla2d nabla^2 d
+        # (see Eq. 1 of https://arxiv.org/abs/2101.12187).
+        #
+        # The BACCO table below contains the following power spectra
+        # in order:
+        # <1,1>
+        # <1,d>
+        # <1,d^2>
+        # <1,s^2>
+        # <1,nabla^2 d>
+        # <d,d>
+        # <d,d^2>
+        # <d,s^2>
+        # <d,nabla^2 d>
+        # <d^2,d^2> (!)
+        # <d^2,s^2> (!)
+        # <d^2,nabla^2 d>
+        # <s^2,s^2> (!)
+        # <s^2,nabla^2 d>
+        # <nabla^2 d,nabla^2 d>
+        #
+        # So:
+        #   a) The spectra involving b2 are for d^2 - convert to d^2/2
+        #   b) The spectra involving bs are for s^2 - convert to s^2/2
+        # Also, the spectra marked with (!) tend to a constant
+        # as k-> 0, which we can suppress with a low-pass filter.
+        #
+        # Importantly, we have corrected the spectra involving s2 to
+        # make the definition of bs equivalent in the EPT and LPT
+        # expansions.
+        bL11 = b11-1
+        bL12 = b12-1
+        if Pnl is None:
+            Pdmdm = self.bacco_table[:, 0, :]
+            Pdmd1 = self.bacco_table[:, 1, :]
+            Pd1d1 = self.bacco_table[:, 5, :]
+            pgg = (Pdmdm + (bL11+bL12)[:, None] * Pdmd1 +
+                   (bL11*bL12)[:, None] * Pd1d1)
+        else:
+            pgg = (b11*b12)[:, None]*Pnl
+        Pdmd2 = 0.5*self.bacco_table[:, 2, :]
+        Pd1d2 = 0.5*self.bacco_table[:, 6, :]
+        Pd2d2 = 0.25*self.bacco_table[:, 9, :]*self.wk_low[None, :]
+        Pdms2 = 0.5*self.bacco_table[:, 3, :]
+        Pd1s2 = 0.5*self.bacco_table[:, 7, :]
+        Pd2s2 = 0.25*self.bacco_table[:, 10, :]*self.wk_low[None, :]
+        Ps2s2 = 0.25*self.bacco_table[:, 12, :]*self.wk_low[None, :]
+        #TODO: OK to use low-pass filter?
```

Comment:
By default we set it to 1 anyway, right?

Reply:
No I don't think so. Looking at https://github.com/LSSTDESC/DESCBiasChallenge/blob/984e61e0c7387988841058f2017a9a3581e604c7/likelihood/cl_like/cl_like.py#L34, I think it is actually applied. But the low-k behavior for HEFT is similar to LPT, so if we apply it there, I think we should also do it for BACCO.